contracts/BC_fusion/StakePool.sol

```diff
@@ -0,0 +1,248 @@
+        require(success, "TRANSFER_FAILED");
+        return _bnbAmount;
+    }
+
+    function claim(address payable _delegator, uint256 number) external onlyStakeHub whenNotPaused returns (uint256) {
+        require(_unbondRequestsQueue[_delegator].length() != 0, "NO_UNBOND_REQUEST");
+        // number == 0 means claim all
+        if (number == 0) {
+            number = _unbondRequestsQueue[_delegator].length();
+        }
+        if (number > _unbondRequestsQueue[_delegator].length()) {
+            number = _unbondRequestsQueue[_delegator].length();
+        }
+        require(number <= MAX_CLAIM_NUMBER, "TOO_MANY_REQUESTS"); // prevent too many loop in one transaction
+
+        uint256 _totalShares;
+        uint256 _totalBnbAmount;
+        while (number != 0) {
+            bytes32 hash = _unbondRequestsQueue[_delegator].front();
+            UnbondRequest memory request = _unbondRequests[hash];
+            if (block.timestamp < request.unlockTime) {
+                break;
+            }
+
+            // request is non-existed(should not happen)
+            if (request.sharesAmount == 0 && request.unlockTime == 0) {
+                continue;
+            }
+
+            _totalShares += request.sharesAmount;
+            _totalBnbAmount += request.bnbAmount;
+
+            // remove from the queue
+            _unbondRequestsQueue[_delegator].popFront();
+            delete _unbondRequests[hash];
+
+            number -= 1;
+        }
+
+        _lockedShares[_delegator] -= _totalShares;
+        (bool success,) = _delegator.call{value: _totalBnbAmount}("");
+        require(success, "CLAIM_FAILED");
+
+        emit UnbondClaimed(_delegator, _totalShares, _totalBnbAmount);
+        return _totalBnbAmount;
+    }
+
+    function distributeReward() external payable onlyStakeHub {
+        uint256 _bnbAmount = msg.value;
+        _totalReceivedReward += _bnbAmount;
+        _totalPooledBNB += _bnbAmount;
+        emit RewardReceived(_bnbAmount);
+    }
+
+    function slash(uint256 _slashBnbAmount) external onlyStakeHub returns (uint256) {
+        uint256 selfDelegation = _sharesOf(validator);
+        uint256 _slashShares = getSharesByPooledBNB(_slashBnbAmount);
+
+        uint256 _remainingSlashBnbAmount = _slashBnbAmount;
+        uint256 _remainingSlashShares = _slashShares;
+        if (_slashShares <= selfDelegation) {
+            _totalPooledBNB -= _slashBnbAmount;
+            _burn(validator, _slashShares);
+            _remainingSlashBnbAmount = 0;
+        } else {
+            uint256 selfDelegationTokens = getPooledBNBByShares(selfDelegation);
+            _totalPooledBNB -= selfDelegationTokens;
+            _burn(validator, selfDelegation);
+
+            _remainingSlashBnbAmount -= selfDelegationTokens;
+            _remainingSlashShares -= selfDelegation;
+
+            uint256 _unbondingShares = _lockedShares[validator];
+            while (_remainingSlashBnbAmount > 0 && _unbondingShares > 0) {
```

Comment:
It is possible a DOS attack and never end.

Reply:
Done